synthtool/gcp/templates/ruby_library/AUTHENTICATION.md

```diff
@@ -0,0 +1,179 @@
+# Authentication
+
+In general, the {{ metadata["package_name"] }} library uses [Service
+Account](https://cloud.google.com/iam/docs/creating-managing-service-accounts)
+credentials to connect to Google Cloud services. When running on Compute Engine
+the credentials will be discovered automatically. When running on other
+environments, the Service Account credentials can be specified by providing the
+path to the [JSON
+keyfile](https://cloud.google.com/iam/docs/managing-service-account-keys) for
+the account (or the JSON itself) in environment variables. Additionally, Cloud
+SDK credentials can also be discovered automatically, but this is only
+recommended during development.
+
+## Project and Credential Lookup
+
+The {{ metadata["package_name"] }} library aims to make authentication
+as simple as possible, and provides several mechanisms to configure your system
+without providing **Project ID** and **Service Account Credentials** directly in
+code.
+
+**Project ID** is discovered in the following order:
+
+1. Specify project ID in method arguments
+2. Specify project ID in configuration
+3. Discover project ID in environment variables
+4. Discover GCE project ID
```

Comment:
Dunno how in-depth you want to get, but if no project_id is provided, credentials will also try to extract project_id from json before checking `gcloud config config-helper --format json`

Reply:
Our JSON files were created before `project_id` was added to them, so I think we forget this too often. Most of the time, users can specify _only_ credentials and it will work.

Reply:
Added:

> 5. Discover project ID in credentials JSON

Does that sound OK?